### ISSUE TITLE
Bug: List shadow DOM slot projection breaks query-based child discovery

### ISSUE THREAD

**Issue by driskull** [Esri/calcite-components#14855]
**monday.com sync:** #12594191617

# Bug Report Draft: Shadow DOM slot projection breaks query-based child discovery

## Check existing issues
- [x] I have checked for existing issues to avoid duplicates

## Which Component
- List

## Actual Behavior
calcite-list tracks items/groups with querySelectorAll. Projected list item structures can desync selection, filtering, sorting, and keyboard focus state.

## Expected Behavior
The component should track slotted child items via slot-based APIs (for example, slotchange + assignedElements with flatten as needed), so behavior remains correct when children are projected through parent Shadow DOM.

## Reproduction Sample
https://codepen.io/pen?template=emzLWmy

## Reproduction Version
5.2.0-next.29

## Relevant Info
- OS: macOS
- Browsers: Latest Chrome and Safari
- Root cause pattern: querySelector/querySelectorAll-based child discovery misses projected children when crossing slot boundaries.

## Regression?
No. Not a regression.

## Priority impact
impact - p2 - want for an upcoming milestone

## Impact
Breaks reliability for wrapper-component architectures that compose Calcite components through Shadow DOM projection.

## Calcite package
- [x] @esri/calcite-components
- [ ] @esri/calcite-components-react
- [ ] @esri/calcite-design-tokens
- [ ] @esri/calcite-ui-icons
- [ ] @esri/eslint-plugin-calcite-components

## Esri team
N/A

## Reproduction Steps
1. Paste sample and select items.
2. Navigate with Arrow keys.
3. Open console and verify projected list-item count.
4. Observe selectedItems/focus behavior can miss projected entries.

## CodePen Failure Example
Paste the following into the CodePen template to reproduce the issue pattern:

```html
<wrapper-host id="host">
  <calcite-list-item label="Alpha" value="a"></calcite-list-item>
  <calcite-list-item label="Beta" value="b"></calcite-list-item>
  <calcite-list-item label="Gamma" value="c"></calcite-list-item>
</wrapper-host>

<script type="module">
  customElements.define("wrapper-host", class extends HTMLElement {
    constructor() {
      super();
      const root = this.attachShadow({ mode: "open" });
      root.innerHTML = `
        <calcite-list id="list" selection-mode="single">
          <slot></slot>
        </calcite-list>
      `;
    }
  });

  const host = document.querySelector("#host");
  console.log("Projected list items:", host.querySelectorAll("calcite-list-item").length);
</script>
```

**Comment by annelfitz** [comment]
for some context - this is affecting the keyboard accessibility of the search result renderer component in the JavaScript Maps SDK

**Comment by eriklharper** [comment]
Why doesn't the root element just accept a list in light dom like this?  Why render the list element in the shadow root?
```
<wrapper-host id="host">
  <calcite-list>
    <calcite-list-item label="Alpha" value="a"></calcite-list-item>
    <calcite-list-item label="Beta" value="b"></calcite-list-item>
    <calcite-list-item label="Gamma" value="c"></calcite-list-item>
  </calcite-list?
</wrapper-host>

**Comment by driskull** [comment]
> Why doesn't the root element just render a list in light dom like this? Why render the list element in the shadow root?

@eriklharper Not sure I understand. Some components may want to render a list within their shadow DOM and I believe that should be supported.

**Comment by eriklharper** [comment]
>Some components may want to render a list within their shadow DOM and I believe that should be supported.

I understand that shadow dom components would want to render a list within their shadow DOM, but that's not how your example shows it.  It renders just the `calcite-list` in the shadow dom, but the items are light dom.  What I'm asking is why wouldn't the outer component just accept as light dom the outer `calcite-list` element in addition to the items?

Sounds like the reason is so the outer component can control the properties on the `calcite-list` element, but I haven't seen the context where this would apply, so that's probably why it looks strange to me.  Maybe I just need to learn more about the real world use case.

**Comment by driskull** [comment]
> What I'm asking is why wouldn't the outer component just accept as light dom the outer calcite-list element in addition to the items?

Because in a lot of newer use cases users want to slot in items and not the whole component. For example, if a user is already rendering a list within the shadow dom of their component but they want to allow users to slot in extra list items. This is widely done in other components like action-bars where users can slot in extra actions along with the existing ones.

Otherwise, we would be saying that all parent/child relationships would need to be within the same shadowDom. Maybe we can say that for some components? Its a tough sell for things like actions where users want to slot them into larger components where they are internally placed into an action bar. cc @jcfranco

**Comment by eriklharper** [comment]
> > What I'm asking is why wouldn't the outer component just accept as light dom the outer calcite-list element in addition to the items?
> 
> Because in a lot of newer use cases users want to slot in items and not the whole component. For example, if a user is already rendering a list within the shadow dom of their component but they want to allow users to slot in extra list items. This is widely done in other components like action-bars where users can slot in extra actions along with the existing ones.
> 
> Otherwise, we would be saying that all parent/child relationships would need to be within the same shadowDom. Maybe we can say that for some components? Its a tough sell for things like actions where users want to slot them into larger components where they are internally placed into an action bar. cc [@jcfranco](https://github.com/jcfranco)

Thanks for the explainer!  I like to understand the context, and that helped, so thank you!

**Comment by driskull** [comment]
No problem! Maybe this is overkill for all components and we should only do it for ones that have a good use case. So its good to discuss this and get confirmation from @jcfranco 👍

**Comment by jcfranco** [comment]
I think the use case makes sense. With `action-bar` and `flow` (soon), it looks like this pattern is emerging more, so it'd be helpful to have a wiki page entry covering the use case/pattern for future reference. Per our chat, @driskull will provide a draft.

**Comment by driskull** [comment]
@jcfranco @eriklharper I added a pattern here: https://github.com/Esri/calcite-design-system/wiki/Coding-Conventions#shadow-dom-slot-projection-and-child-discovery

Feel free to edit as necessary
